File: apps/agent/src/agent/tools/get-briefing-context.ts
import { embedQuery } from "../../lib/vertex.js";
import { config } from "../../config.js";
import { getCollections } from "../../lib/mongo.js";
import type { Citation, ToolDef } from "../types.js";

export const getBriefingContextTool: ToolDef = {
  declaration: {
    name: "get_briefing_context",
    description:
      "Compose context for a 1-pager briefing on a calendar event: attendees, open threads with them, outstanding commitments, and recent artifacts referenced. Use the event title or id.",
    parameters: {
      type: "OBJECT",
      properties: {
        event_title: {
          type: "string",
          description: "Title of the calendar event. Case-insensitive substring match is used.",
        },
        event_id: {
          type: "string",
          description: "Optional Mongo _id of the calendar event document.",
        },
      },
    },
  },
  handler: async (args) => {
    try {
      const { documents, chunks } = await getCollections();
      const eventTitle =
        typeof args["event_title"] === "string" ? args["event_title"].trim() : "";
      if (!eventTitle && !args["event_id"]) {
        return { ok: false, error: "either event_title or event_id is required" };
      }

      const event = eventTitle
        ? await documents.findOne({
            source: "calendar",
            title: { $regex: eventTitle, $options: "i" },
          })
        : null;
      if (!event) {
        return {
          ok: false,
          error: `no calendar event matching "${eventTitle}"`,
        };
      }

      const attendees = (event.metadata?.["attendees"] as string[]) ?? [];
      const threadId = event.metadata?.["threadId"] as string | undefined;

      const vector = await embedQuery(
        `${event.title}\n\n${typeof event.body === "string" ? event.body : ""}`,
      );
      const related = await chunks
        .aggregate<Record<string, unknown>>([
          {
            $vectorSearch: {
              index: config.MONGODB_VECTOR_INDEX,
              path: "embedding",
              queryVector: vector,
              numCandidates: 200,
              limit: 8,
            },
          },
          {
            $project: {
              _id: 1,
              documentId: 1,
              source: 1,
              title: 1,
              text: 1,
              ordinal: 1,
              score: { $meta: "vectorSearchScore" },
              metadata: 1,
            },
          },
        ])
        .toArray();

      const commitmentChunks = await chunks
        .find(
          threadId
            ? { "metadata.threadId": threadId }
            : {
                text: {
                  $regex:
                    "(owe|owed|will deliver|by Friday|by Mon|by Tue|by Wed|by Thu|by EOD|by next|committed|promised)",
                  $options: "i",
                },
              },
          { limit: 12, projection: { _id: 1, title: 1, source: 1, text: 1, metadata: 1 } },
        )
        .toArray();

      const citations: Citation[] = related.map((r) => ({
        chunkId: String(r["_id"]),
        documentId: String(r["documentId"]),
        source: r["source"] as Citation["source"],
        title: String(r["title"]),
        score: Number(r["score"] ?? 0),
        ordinal: Number(r["ordinal"] ?? 0),
      }));

      return {
        ok: true,
        data: {
          event: {
            id: String(event._id),
            title: event.title,
            when: event.metadata?.["eventTime"] ?? event.metadata?.["date"],
            location: event.metadata?.["eventLocation"] ?? null,
            organizer: event.metadata?.["organizer"] ?? null,
            attendees,
            agendaExcerpt: typeof event.body === "string" ? event.body.slice(0, 360) : null,
          },
          related: related.map((r) => ({
            chunkId: String(r["_id"]),
            title: r["title"],
            source: r["source"],
            score: r["score"],
            text: typeof r["text"] === "string" ? (r["text"] as string).slice(0, 280) : null,
          })),
          commitmentLeads: commitmentChunks.map((c) => ({
            chunkId: String(c._id),
            title: c.title,
            source: c.source,
            excerpt: typeof c.text === "string" ? c.text.slice(0, 280) : null,
          })),
        },
        citations,
        summary: `briefing pulled for "${event.title}" — ${related.length} related, ${commitmentChunks.length} commitment leads`,
      };
    } catch (err) {
      const msg = err instanceof Error ? err.message : String(err);
      return { ok: false, error: msg };
    }
  },
};

File: apps/agent-py/app/agent/tools/get_briefing_context.py
"""get_briefing_context — port of apps/agent/src/agent/tools/get-briefing-context.ts."""
from __future__ import annotations

from app.agent.types import ToolDef
from app.config import settings
from app.db.mongo import chunks, documents
from app.llm.genai_client import embed_query

_DECL = {
    "name": "get_briefing_context",
    "description": (
        "Compose context for a 1-pager briefing on a calendar event: attendees, open threads with them, "
        "outstanding commitments, and recent artifacts referenced. Use the event title or id."
    ),
    "parameters": {
        "type": "OBJECT",
        "properties": {
            "event_title": {"type": "string", "description": "Title of the calendar event (substring match)."},
            "event_id": {"type": "string", "description": "Optional Mongo _id of the calendar event document."},
        },
    },
}

_COMMIT_RX = r"(owe|owed|will deliver|by Friday|by Mon|by Tue|by Wed|by Thu|by EOD|by next|committed|promised)"


async def _handler(args: dict, ctx: dict | None = None) -> dict:
    try:
        event_title = args["event_title"].strip() if isinstance(args.get("event_title"), str) else ""
        if not event_title and not args.get("event_id"):
            return {"ok": False, "error": "either event_title or event_id is required"}

        event = await documents().find_one({"source": "calendar", "title": {"$regex": event_title, "$options": "i"}}) if event_title else None
        if not event:
            return {"ok": False, "error": f'no calendar event matching "{event_title}"'}

        md = event.get("metadata") or {}
        attendees = md.get("attendees") or []
        thread_id = md.get("threadId")

        vector = await embed_query(f"{event['title']}\n\n{event.get('body') or ''}")
        related = await chunks().aggregate([
            {"$vectorSearch": {"index": settings.mongodb_vector_index, "path": "embedding",
                               "queryVector": vector, "numCandidates": 200, "limit": 8}},
            {"$project": {"_id": 1, "documentId": 1, "source": 1, "title": 1, "text": 1, "ordinal": 1,
                          "score": {"$meta": "vectorSearchScore"}, "metadata": 1}},
        ]).to_list(length=None)

        commit_flt = ({"metadata.threadId": thread_id} if thread_id
                      else {"text": {"$regex": _COMMIT_RX, "$options": "i"}})
        commit_chunks = await chunks().find(
            commit_flt, projection={"_id": 1, "title": 1, "source": 1, "text": 1, "metadata": 1}, limit=12
        ).to_list(length=None)

        citations = [{
            "chunkId": str(r["_id"]), "documentId": str(r.get("documentId", "")), "source": r.get("source"),
            "title": str(r.get("title", "")), "score": float(r.get("score", 0)), "ordinal": int(r.get("ordinal", 0)),
        } for r in related]

        return {"ok": True, "data": {
            "event": {
                "id": str(event["_id"]), "title": event["title"],
                "when": md.get("eventTime") or md.get("date"), "location": md.get("eventLocation"),
                "organizer": md.get("organizer"), "attendees": attendees,
                "agendaExcerpt": (event.get("body") or "")[:360] if isinstance(event.get("body"), str) else None,
            },
            "related": [{"chunkId": str(r["_id"]), "title": r.get("title"), "source": r.get("source"),
                         "score": r.get("score"), "text": (r.get("text") or "")[:280]} for r in related],
            "commitmentLeads": [{"chunkId": str(c["_id"]), "title": c.get("title"), "source": c.get("source"),
                                 "excerpt": (c.get("text") or "")[:280]} for c in commit_chunks],
        }, "citations": citations,
            "summary": f'briefing pulled for "{event["title"]}" — {len(related)} related, {len(commit_chunks)} commitment leads'}
    except Exception as err:  # noqa: BLE001
        return {"ok": False, "error": str(err)}


tool = ToolDef(declaration=_DECL, handler=_handler)
